src/misc/cross.cpp

```diff
@@ -715,6 +715,88 @@ std::string cfstr_to_string(CFStringRef source)
 	return lang;
 }
 
+#if defined (__linux__)
+
+#include <fcntl.h>
+
+int get_num_physical_cpus()
+{
+	constexpr int MaxCpus = 128;
+	bool core_ids[MaxCpus] = {};
+	int num_cores = 0;
+	for (int i = 0; i < MaxCpus; ++i) {
+		std::string path = "/sys/devices/system/cpu/cpu" + std::to_string(i) + "/topology/core_id";
+		int fd = open(path.c_str(), O_RDONLY);
+		if (fd != -1) {
+			char core_id_str[6] = {};
+			if (read(fd, core_id_str, 5) > 0) {
+				int core_id = atoi(core_id_str);
+				if (core_id >= 0 && core_id < MaxCpus && !core_ids[core_id]) {
+					core_ids[core_id] = true;
+					++num_cores;
+				}
+			}
+			close(fd);
+		}
+	}
+	return std::max(num_cores, 1);
+}
+
+#elif defined (MACOSX)
+
+#include <sys/sysctl.h>
+
+int get_num_physical_cpus()
+{
+	int num_cpus = 0;
+	size_t size = sizeof(num_cpus);
+	sysctlbyname("hw.physicalcpu", &num_cpus, &size, nullptr, 0);
+	return std::max(num_cpus, 1);
+}
+
+#elif defined (WIN32)
+
+#include <Windows.h>
+
+int get_num_physical_cpus()
+{
+	// Size of the buffer as input, gets modified on function call to size of return data
+	DWORD buffer_size = 10 * 1024 * 1024;
+	uint8_t* cpu_info = (uint8_t *)malloc(buffer_size);
+	GetLogicalProcessorInformationEx(RelationProcessorCore,
+	                                 (PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX)cpu_info,
+	                                 &buffer_size);
+
+	// Return value is a variably sized struct and requires type-punning to read
+	// Whoever made this API was probably on substances that are illegal in many countries
+	DWORD bytes_read = 0;
+	int num_entries  = 0;
+	while (bytes_read < buffer_size) {
+		PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX info =
+		        (PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX)(cpu_info +
+		                                                   bytes_read);
+		bytes_read += info->Size;
+		++num_entries;
+	}
+	free(cpu_info);
+
+	// There should be one entry per core so all we care about is the number of entries we got back
+	return std::max(num_entries, 1);
```

Comment:
🧿🌈🍄☮🪴 substance-induced API, indeed! Wow.

I went back to the `GetLogicalProcessorInformation` (win 2003+), versus the newer `GetLogicalProcessorInformationEx` (win 7+), and it looks like the prior non-Ex API is slightly more sane; probably when they were younger and smarter :grimacing: 

It could be used something like this:

```c++

DWORD total_bytes = 0;
if (!GetLogicalProcessorInformation(nullptr, &total_bytes)) {
    return 1;
}

const auto num_infos = total_bytes / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION);

std::vector<SYSTEM_LOGICAL_PROCESSOR_INFORMATION> infos(num_infos);
    
if (!GetLogicalProcessorInformation(infos.data(), &total_bytes)) {
    return 1;
}

int num_hw_cores = 0;
for(const auto& info : infos) {
    if (info.Relationship == RelationProcessorCore) {
        ++num_hw_cores;
    }
}

return std::max(num_hw_cores, 1);
```

Reply:
> I went back to the `GetLogicalProcessorInformation` (win 2003+), versus the newer `GetLogicalProcessorInformationEx` (win 7+), and it looks like the prior non-Ex API is slightly more sane; probably when they were younger and smarter 😬

Those psychedelics take a heavy tool if you keep mixing them for decades. Just look at the seminal final works of Philip K. Dick (I love him, by the way, but man, half of those ideas must be the result of some _heavy duty_ substance abuse, I have zero doubts about that... 😎)

Reply:
Sounds like I should read some of his stuff :smile: (At least Philip wasn't designing OS APIs! :laughing:)

Reply:
Well, the film industry has been profiting from adapting his books to movies for quite some time:
https://en.wikipedia.org/wiki/List_of_adaptations_of_works_by_Philip_K._Dick

The mini-series Electric Dreams was also quite good:
https://en.wikipedia.org/wiki/Electric_Dreams_(2017_TV_series)

I quite enjoyed the novels Ubik, VALIS, The Three Stigmata of Palmer Eldritch, Martian Timeslip, Flow My Tears the Policeman Said, and The Transmigration of Timothy Archer.

Reply:
@johnnovak, thanks for the tips to get me going! I'm a newbie to his work; looking forward to it 🤘 🍻

Reply:
I think I'll stick to the Ex version.  It works and it's more flexible which we may end up needing if we want to identify E cores in the future.

Good example though.  I added a missing error check thanks to that.